src/main/java/io/harness/cf/client/api/MetricsProcessor.java

```diff
@@ -39,42 +34,95 @@
+  private final MetricsApi api;
+
+  private String jarVersion = "";
+
+  @Setter private String environmentID;
+  @Setter private String cluster;
 
-  public AnalyticsPublisherService(
-      MetricsApi metricsAPI,
-      Config config,
-      String environmentID,
-      String cluster,
-      Cache analyticsCache) {
-
-    this.metricsAPI = metricsAPI;
-    this.analyticsCache = analyticsCache;
-    this.environmentID = environmentID;
-    this.cluster = cluster;
+  private String token;
+
+  public MetricsProcessor(Config config, MetricsCallback callback) {
     this.config = config;
+    this.callback = callback;
+    this.api = new MetricsApi();
+    this.queue = new LinkedBlockingQueue<>(config.getBufferSize());
+    this.callback.onMetricsReady();
+  }
+
+  public void setToken(String token) {
+    this.token = token;
+    this.api.setApiClient(makeApiClient());
+  }
+
+  protected ApiClient makeApiClient() {
+    final int maxTimeout = 30 * 60 * 1000;
+    ApiClient apiClient = new ApiClient();
+    apiClient.setBasePath(config.getEventUrl());
+    apiClient.setConnectTimeout(maxTimeout);
+    apiClient.setReadTimeout(maxTimeout);
+    apiClient.setWriteTimeout(maxTimeout);
+    apiClient.setDebugging(log.isDebugEnabled());
+    apiClient.addDefaultHeader("Authorization", "Bearer " + token);
+    return apiClient;
+  }
+
+  // push the incoming data to the ring buffer
+  public void pushToQueue(Target target, FeatureConfig featureConfig, Variation variation) {
+
+    MetricEvent analytics =
+        MetricEvent.builder()
+            .featureConfig(featureConfig)
+            .target(target)
+            .variation(variation)
+            .build();
+
+    if (queue.remainingCapacity() == 0) {
+      List<MetricEvent> data = new ArrayList<>();
+      queue.drainTo(data);
+      sendDataAndResetCache(data);
+    }
+
+    try {
+      MetricEvent event = new MetricEvent();
+      event.setFeatureConfig(analytics.getFeatureConfig());
+      event.setTarget(analytics.getTarget());
+      event.setVariation(analytics.getVariation());
+      queue.put(event);
+    } catch (InterruptedException e) {
+
+      log.debug("Long waiting");
+    }
   }
 
   /** This method sends the metrics data to the analytics server and resets the cache */
-  public void sendDataAndResetCache() throws CfClientException {
+  public void sendDataAndResetCache(final List<MetricEvent> data) {
     log.debug("Reading from queue and building cache");
-    jarVerion = getVersion();
+    jarVersion = getVersion();
+
+    if (!data.isEmpty()) {
 
-    final Map<Analytics, Integer> all = analyticsCache.getAll();
-    if (!all.isEmpty()) {
       try {
+        Map<MetricEvent, Integer> map = new HashMap<>();
+        for (MetricEvent event : data) {
+          Integer value = map.get(event);
+          if (value != null) {
+            value += 1;
+            map.put(event, value);
+          }
+          map.put(event, 0);
+        }
```

Comment:
could be easier with streams (also works on empty collection). E.g. something like `data.collect(Collectors.groupingBy(Function.identity(), Collectors.counting()));`